### PR TITLE
Add docs for ACME setup

### PR DESCRIPTION
**Pull request by anoadragon453**
Add documentation for ACME certificate provisioning in Synapse v0.99.0+.

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -227,6 +227,85 @@ to read `Using a reverse proxy with Synapse`_ when doing so.
...
+ACME setup
+----------
+
+Synapse requires valid TLS certificates for communication between servers
+(port ``8448`` by default) in addition to those that are client-facing (port
+``443``). Synapse **will provision server-to-server certificates
+automatically for you for free** through `Let's Encrypt
+<https://letsencrypt.org/>`_ if you tell it to.
+
+    Note: Synapse does not currently hot-renew Let's Encrypt certificates for
```
Comment (richvdh) [inline README.rst]: meh, let's leave this out

File: README.rst [hunk 2]
```diff
@@ -227,6 +227,85 @@ to read `Using a reverse proxy with Synapse`_ when doing so.
 Apart from port 8448 using TLS, both ports are the same in the default
 configuration.
 
+ACME setup
+----------
+
+Synapse requires valid TLS certificates for communication between servers
+(port ``8448`` by default) in addition to those that are client-facing (port
+``443``). Synapse **will provision server-to-server certificates
```
Comment (richvdh) [inline README.rst]: we should mention that they can be the same cert if the `server_name` is the same as the hostname that the client connects to.

File: README.rst [hunk 3]
```diff
@@ -227,6 +227,85 @@ to read `Using a reverse proxy with Synapse`_ when doing so.
...
+
+In order for Synapse to complete the ACME challenge to provision a
+certificate, it needs access to port 80. Typically listening on port 80 is
+only granted to applications running as root. There are thus two solutions to
+this problem.
+
+**Using a reverse proxy**
+
+A reverse proxy such as Apache or Nginx allows a single process (the web
+server) to listen on port 80 and redirect traffic to the appropriate program
```
Comment (richvdh) [inline README.rst]: s/redirect/proxy/

File: README.rst [hunk 4]
```diff
@@ -227,6 +227,85 @@ to read `Using a reverse proxy with Synapse`_ when doing so.
...
+
+``authbind`` allows a program which does not or should not run as root to
+bind to low-numbered ports in a controlled way. The setup is simpler, but
+requires a webserver not to already be running on port 80. **This includes
+every time Synapse renews a certificate**, which may be cumbersome if you
+usually run a web server on port 80. Nevertheless, if you're sure port 80 is
+not being used for any other purpose then all that is necessary is the
+following:
+
+Install ``authbind``::
```
Comment (richvdh) [inline README.rst]:
```suggestion
Install ``authbind``. For example, on Debian/Ubuntu::
```

File: README.rst [hunk 5]
```diff
@@ -227,6 +227,85 @@ to read `Using a reverse proxy with Synapse`_ when doing so.
...
+
+    sudo touch /etc/authbind/byport/80
+    sudo chmod 777 /etc/authbind/byport/80
+
+When Synapse is started (do not start it yet), use the following syntax::
+
+    # authbind syntax. don't start Synapse yet
+    authbind --deep <synapse start command>
+
+If using the `Systemd`_ service file above, you can change the following line
```
Comment (richvdh) [inline README.rst]: there is no systemd file; need to change this to use synctl

File: README.rst [hunk 6]
```diff
@@ -212,20 +212,87 @@ key in the ``<server name>.signing.key`` file (the second word) to something
...
-you can temporarly use another certificate. (Note that a self-signed
-certificate is fine for `Federation`_). You can do so by changing
-``tls_certificate_path`` and ``tls_private_key_path``
-in ``homeserver.yaml``; alternatively, you can use a reverse-proxy, but be sure
-to read `Using a reverse proxy with Synapse`_ when doing so.
+is configured to use TLS for `Federation`_ with a self-signed or verified
+certificate, but please be aware that a valid certificate will be required in
+Synapse v1.0.
+
+If you would like to do initial testing with a client without having to setup
```
Comment (richvdh) [inline README.rst]: I don't think this is limited to initial testing?

File: README.rst [hunk 7]
```diff
@@ -553,9 +621,7 @@ you to run your server on a machine that might not have the same name as your
 domain name. For example, you might want to run your server at
 ``synapse.example.com``, but have your Matrix user-ids look like
 ``@user:example.com``. (A SRV record also allows you to change the port from
-the default 8448. However, if you are thinking of using a reverse-proxy on the
-federation port, which is not recommended, be sure to read
-`Reverse-proxying the federation port`_ first.)
+the default 8448.
```
Comment (richvdh) [inline README.rst]:
```suggestion
the default 8448.)
```

File: README.rst [hunk 8]
```diff
@@ -705,63 +767,7 @@ Having done so, you can then use ``https://matrix.example.com`` (instead of
...
-
-The most common mistake when setting up federation is not to tell Synapse about
-your SSL certificate. To check it, you can visit
-``https://matrix.org/federationtester/api/report?server_name=<your_server_name>``.
-Unfortunately, there is no UI for this yet, but, you should see
-``"MatchingTLSFingerprint": true``. If not, check that
-``Certificates[0].SHA256Fingerprint`` (the fingerprint of the certificate
-presented by your reverse-proxy) matches ``Keys.tls_fingerprints[0].sha256``
-(the fingerprint of the certificate Synapse is using).
+Please see `ACME setup`_ for details on reverse-proxying the federation port.
```
Comment (richvdh) [inline README.rst]: the acme setup section doesn't say anything about reverse-proxying port 8448. Suggest removing this sentence.

File: README.rst [hunk 9]
```diff
@@ -216,16 +216,77 @@ different. See `the spec`__ for more information on key management.)
...
+certificate, but please be aware that a valid certificate will be required in
+Synapse v1.0.
 
-Apart from port 8448 using TLS, both ports are the same in the default
-configuration.
+
+ACME setup
+----------
+
+Synapse v1.0 requires valid TLS certificates for communication between servers
```
Comment (richvdh) [inline README.rst]:
```suggestion
Synapse v1.0 will require valid TLS certificates for communication between servers
```

File: README.rst [hunk 10]
```diff
@@ -657,14 +717,10 @@ port. Indeed, clients will use port 443 by default, whereas servers default to
...
-The next most important thing to know is that using a reverse-proxy on the
-federation port has a number of pitfalls. It is possible, but be sure to read
-`Reverse-proxying the federation port`_.
-
-The recommended setup is therefore to configure your reverse-proxy on port 443
-to port 8008 of synapse for client connections, but to also directly expose port
-8448 for server-server connections. All the Matrix endpoints begin ``/_matrix``,
-so an example nginx configuration might look like::
+The recommended setup is therefore to configure your reverse-proxy on port
+443 to port 8008 of synapse for client connections, and port 8448 for
```
Comment (richvdh) [inline README.rst]: this makes no sense.

I think the recommendation (if you're going to use a reverse-proxy) should be:

* set `no_tls: True` and remove the `tls` listeners from the synapse config
* have reverse proxy listen on both 443 and 8448, and forward both to port 8008.

However, this is starting to all get out of scope. For now I suggest: you remove this paragraph and just say "An example nginx configuration for forwarding client connections to synapse is::"

File: README.rst [hunk 11]
```diff
@@ -216,16 +216,77 @@ different. See `the spec`__ for more information on key management.)
...
+
+When Synapse is started, use the following syntax::
+
+    authbind --deep <synapse start command>
+
+If you would like to use your own certificates, you can do so by
+changing ``tls_certificate_path`` and ``tls_private_key_path`` in
+``homeserver.yaml``; alternatively, you can use a reverse-proxy, but be sure
+to read `Using a reverse proxy with Synapse`_ when doing so. Apart from port
+8448 using TLS, both ports are the same in the default configuration.
```
[the remark this answers is not in the document]
Comment (anoadragon453) [reply README.rst]: I'm not sure what this is talking about. Tempted to just take out the line.

**Pull request merged**